File: src/main/java/com/letgo/book/application/create/CreateBookCommandHandler.java
package com.letgo.book.application.create;

import com.letgo.book.domain.Book;
import com.letgo.book.domain.BookId;
import com.letgo.book.domain.BookRepository;
import com.letgo.book.domain.BookTitle;
import com.letgo.shared.application.bus.command.CommandHandler;
import com.letgo.shared.application.event.DomainEventPublisher;

final public class CreateBookCommandHandler implements CommandHandler<CreateBookCommand> {
    private final BookRepository repository;
    private final DomainEventPublisher publisher;

    public CreateBookCommandHandler(BookRepository repository, DomainEventPublisher publisher) {
        this.repository = repository;
        this.publisher = publisher;
    }

    @Override
    public void handle(CreateBookCommand command) {
        Book book = Book.create(BookId.create(command.id()), BookTitle.create(command.title()));
        repository.save(book);
        publisher.publish(book.retrieveEvents());
    }
}

File: src/main/java/com/letgo/book/application/create/CreateBookCommandHandler.kt
package com.letgo.book.application.create

import com.letgo.book.domain.Book
import com.letgo.book.domain.BookId
import com.letgo.book.domain.BookRepository
import com.letgo.shared.application.bus.command.CommandHandler
import com.letgo.shared.application.bus.event.DomainEventPublisher

class CreateBookCommandHandler(
    private val repository: BookRepository,
    private val publisher: DomainEventPublisher
) : CommandHandler<CreateBookCommand> {
    override fun handle(command: CreateBookCommand) {
        repository.find(BookId.create(command.id)) ?: Book.create(command.id, command.title, command.occurredOn).also {
            repository.save(it)
            publisher.publish(it.retrieveEvents())
        }
    }
}
